lib/config_cat/config/user_comparator.ex

```diff
@@ -0,0 +1,539 @@
+    end
+  end
+
+  defp compare_datetimes(user_value, comparison_value, valid_comparisons) do
+    with {:ok, user_seconds} <- to_unix_seconds(user_value),
+         {:ok, comparison_seconds} <- to_float(comparison_value) do
+      result =
+        cond do
+          user_seconds < comparison_seconds -> :lt
+          user_seconds > comparison_seconds -> :gt
+          true -> :eq
+        end
+
+      {:ok, result in valid_comparisons}
+    else
+      {:error, :invalid_float} -> {:error, :invalid_datetime}
+      error -> error
+    end
+  end
+
+  defp hash_value(value, %ComparisonContext{} = context) do
+    salted = value <> context.salt <> context.context_salt
+
+    :sha256
+    |> :crypto.hash(salted)
+    |> Base.encode16()
+    |> String.downcase()
+  end
+
+  defp parse_comparison(value) do
+    [length_string, comparison_string] = String.split(value, "_", parts: 2)
+
+    {String.to_integer(length_string), comparison_string}
+  end
+
+  defp as_text(value, _context) when is_binary(value), do: {:ok, value}
+
+  defp as_text(value, %ComparisonContext{} = context) do
+    %ComparisonContext{condition: condition, key: key} = context
+    attribute_name = UserCondition.comparison_attribute(condition)
+    condition_text = UserCondition.description(condition)
+
+    ConfigCatLogger.warning(
+      "Evaluation of condition (#{condition_text}) for setting '#{key}' may not produce the expected result " <>
+        "(the User.#{attribute_name} attribute is not a string value, thus it was automatically converted to " <>
+        "the string value '#{value}'). Please make sure that using a non-string value was intended.",
+      event_id: 3005
+    )
+
+    user_value_to_string(value)
+  end
+
+  @spec user_value_to_string(Config.value() | DateTime.t() | NaiveDateTime.t() | [String.t()]) ::
+          {:ok, String.t() | nil} | {:error, :invalid_datetime | :invalid_float | :invalid_string_list}
+  def user_value_to_string(nil), do: {:ok, nil}
+
+  def user_value_to_string(%DateTime{} = dt) do
+    with {:ok, seconds} <- to_unix_seconds(dt) do
+      {:ok, to_string(seconds)}
+    end
+  end
+
+  def user_value_to_string(%NaiveDateTime{} = naive) do
+    naive |> DateTime.from_naive!("Etc/UTC") |> user_value_to_string()
+  end
+
+  def user_value_to_string(value) when is_list(value) do
+    with {:ok, list} <- to_string_list(value) do
+      {:ok, to_string(list)}
+    end
+  end
+
+  def user_value_to_string(value), do: {:ok, to_string(value)}
+
+  defp to_float(value) when is_float(value), do: {:ok, value}
+  defp to_float(value) when is_integer(value), do: {:ok, value * 1.0}
+
+  defp to_float(value) when is_binary(value) do
+    value
+    |> String.trim()
+    |> String.replace(",", ".")
+    |> Float.parse()
+    |> case do
+      {float, ""} -> {:ok, float}
+      _ -> {:error, :invalid_float}
+    end
+  end
+
+  defp to_float(_value), do: {:error, :invalid_float}
+
+  defp to_string_list(value) when is_list(value), do: {:ok, value}
```

Comment:
As far as I can tell, this doesn't prevent list items other then strings, right? I think we need the same check here as below.

See also: https://github.com/configcat/common-js/blob/68e23064e70363614c6f8d8bab9619a67fc32646/src/RolloutEvaluator.ts#L781-L788

Reply:
Addressed in 00289e5.